compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -1703,13 +1703,21 @@ fn add_post_link_args(cmd: &mut dyn Linker, sess: &Session, flavor: LinkerFlavor
 /// that are necessary for the linking. They are only present in symbol table but not actually
 /// used in any sections, so the linker will therefore pick relevant rlibs for linking, but
 /// unused `#[no_mangle]` or `#[used]` can still be discard by GC sections.
+///
+/// There's a few internal crates in the standard library (aka libcore and
+/// libstd) which actually have a circular dependence upon one another. This
+/// currently arises through "weak lang items" where libcore requires things
+/// like `rust_begin_unwind` but libstd ends up defining it. To get this
+/// circular dependence to work correctly we declare some of these things
+/// in this synthetic object.
 fn add_linked_symbol_object(
     cmd: &mut dyn Linker,
     sess: &Session,
     tmpdir: &Path,
     symbols: &[(String, SymbolExportKind)],
+    missing_lang_items: &FxHashSet<Symbol>,
```

Comment:
I feel that the logic should be at the caller site instead of here.

Reply:
Or more precisely, these symbols should be added to the `linked_symbols` by modifying the `linked_symbols` function in `linker.rs`.

Reply:
Seems reasonable, we won't have to keep missing lang items in `CrateInfo` then.
(You'll have to pass `used_crates` to `fn linked_symbols` though.)

Reply:
Well, alternatively we can just extend the linked_symbols list in `CrateInfo::new`. Or, make make `add_linked_symbol_object` take an iterator and then pass in chained iterator. I don't have a preferrence in that regard.

I just think that for `add_linked_symbol_object` function, it should really just take a list/set/iterator of symbols and shouldn't be aware of lang items and other stuff.